File: app/schemas/source.py
"""Pydantic schemas for sources."""

import math
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class SourceResponse(BaseModel):
    """Source status response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    source_type: str
    base_url: str
    polling_interval_minutes: int
    is_active: bool
    last_polled_at: datetime | None = None
    created_at: datetime


class SourceHealthResponse(BaseModel):
    """Health check for a specific source."""

    source_id: int
    source_name: str
    is_active: bool
    last_polled_at: datetime | None = None
    last_status: str | None = None
    last_error_message: str | None = None
    alerts_last_24h: int = 0
    error_count_last_24h: int = 0
    unprocessed_count: int = 0


class IngestionLogResponse(BaseModel):
    """Single ingestion log entry."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    status: str
    alerts_fetched: int
    alerts_new: int
    error_message: str | None = None
    started_at: datetime
    completed_at: datetime | None = None


class IngestionLogListResponse(BaseModel):
    """Paginated ingestion log history."""

    items: list[IngestionLogResponse]
    total: int
    page: int
    page_size: int
    pages: int

File: frontend/src/types/source.ts
export interface Source {
  id: number;
  name: string;
  source_type: string;
  base_url: string;
  polling_interval_minutes: number;
  is_active: boolean;
  last_polled_at: string | null;
  created_at: string;
}

export interface SourceHealth {
  source_id: number;
  source_name: string;
  is_active: boolean;
  last_polled_at: string | null;
  last_status: string | null;
  last_error_message: string | null;
  alerts_last_24h: number;
  error_count_last_24h: number;
}

export interface IngestionLog {
  id: number;
  status: string;
  alerts_fetched: number;
  alerts_new: number;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface IngestionLogListResponse {
  items: IngestionLog[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}
